xplique/example_based/projections/base.py

```diff
@@ -0,0 +1,254 @@
+            weights = self.get_weights(projected_inputs, targets)
+            weighted_projected_inputs =  tf.multiply(weights, projected_inputs)
+        return weighted_projected_inputs
+
+    def __call__(
+        self,
+        inputs: Union[tf.Tensor, np.ndarray],
+        targets: Optional[Union[tf.Tensor, np.ndarray]] = None,
+    ):
+        """project alias"""
+        return self.project(inputs, targets)
+
+    def project_dataset(
+        self,
+        cases_dataset: tf.data.Dataset,
+        targets_dataset: Optional[tf.data.Dataset] = None,
+    ) -> Optional[tf.data.Dataset]:
+        """
+        Apply the projection to a dataset through `Dataset.map`
+
+        Parameters
+        ----------
+        cases_dataset
+            Dataset of samples to be projected.
+        targets_dataset
+            Dataset of targets for the samples.
+
+        Returns
+        -------
+        projected_dataset
+            The projected dataset.
+        """
+        if self.requires_targets and targets_dataset is None:
+            warnings.warn(
+                "The projection requires `targets` but `targets_dataset` is not provided. "\
+                +"`targets` will be computed online, assuming a classification setting. "\
+                +"Hence, online `targets` will be the predicted class one-hot-encoding. "\
+                +"If this is not the expected behavior, please provide a `targets_dataset`.")
+
+        if self.mappable:
+            return self._map_project_dataset(cases_dataset, targets_dataset)
+        return self._loop_project_dataset(cases_dataset, targets_dataset)
+
+    def _map_project_dataset(
+        self,
+        cases_dataset: tf.data.Dataset,
+        targets_dataset: Optional[tf.data.Dataset] = None,
+    ) -> Optional[tf.data.Dataset]:
+        """
+        Apply the projection to a dataset through `Dataset.map`
+
+        Parameters
+        ----------
+        cases_dataset
+            Dataset of samples to be projected.
+        targets_dataset
+            Dataset of targets for the samples.
+
+        Returns
+        -------
+        projected_dataset
+            The projected dataset.
+        """
+        # project dataset, note that projection is done at iteration time
+        if targets_dataset is None:
+            projected_cases_dataset = cases_dataset.map(self.project)
+        else:
+            # in case targets are provided, we zip the datasets and project them together
+            projected_cases_dataset = tf.data.Dataset.zip(
+                (cases_dataset, targets_dataset)
+            ).map(self.project)
+
+        return projected_cases_dataset
+
+    def _loop_project_dataset(
+        self,
+        cases_dataset: tf.data.Dataset,
+        targets_dataset: tf.data.Dataset,
+    ) -> tf.data.Dataset:
+        """
+        Apply the projection to a dataset without `Dataset.map`.
+        Because attribution methods create a `tf.data.Dataset` for batching,
```

Comment:
Because some projections are not compatible with a `tf.data.Dataset.map`.
For example, the attribution methods because they create a ...